### PR TITLE
feat: implement to 10 users feature

### PR DESCRIPTION
**Pull request by OgDev-01**
## What type of PR is this? (check all applicable)

- [x] 🍕 Feature
- [ ] 🐛 Bug Fix
- [ ] 📝 Documentation Update
- [ ] 🎨 Style
- [ ] 🧑‍💻 Code Refactor
- [ ] 🔥 Performance Improvements
- [ ] ✅ Test
- [ ] 🤖 Build
- [ ] 🔁 CI
- [ ] 📦 Chore (Release)
- [ ] ⏩ Revert

## Description
[PR Description]

This pull request implements the "Top 10 Users" feature. It adds the necessary functionality to display a list of the top 10 users based on certain criteria. This feature will provide valuable insights into user activity and engagement. The implementation is complete and ready for review.

### Example response:
image

_Generated using [OpenSauced](https://opensauced.ai/)._

## Related Tickets & Documents
closes https://github.com/open-sauced/ops-backend/issues/76

## Mobile & Desktop Screenshots/Recordings

image

## Added tests?

- [ ] 👍 yes
- [x] 🙅 no, because they aren't needed
- [ ] 🙋 no, because I need help

## Added to documentation?

- [ ] 📜 README.md
- [ ] 📓 docs.opensauced.pizza
- [ ] 🍕 dev.to/opensauced
- [ ] 📕 storybook
- [x] 🙅 no documentation needed

### CODE REVIEW

File: src/highlight/highlight.controller.ts [hunk 1]
```diff
@@ -18,11 +18,18 @@ import { PageDto } from "../common/dtos/page.dto";
 import { DbUserHighlight } from "../user/entities/user-highlight.entity";
 import { DbUserHighlightRepo } from "./entities/user-highlight-repo.entity";
 import { UserHighlightsService } from "../user/user-highlights.service";
+import { UserService } from "../user/user.service";
+// import { DbUser } from "../user/user.entity";
```
Comment (brandonroberts) [inline src/highlight/highlight.controller.ts]:
```suggestion
```

**Comment by OgDev-01** [comment]
I'm not sure I got this right 🫠🫠. I'm open to criticism on this one 😄
cc: @brandonroberts

File: src/highlight/highlight.controller.ts [hunk 2]
```diff
@@ -46,6 +53,16 @@ export class HighlightController {
     return this.userHighlightsService.findAllFeatured(pageOptionsDto);
   }
 
+  @Get("/top-10-users")
```
Comment (brandonroberts) [inline src/highlight/highlight.controller.ts:54]: This endpoint should go in the `users` controller as `/users/top`

File: src/user/user.service.ts [hunk 3]
```diff
@@ -23,6 +27,28 @@ export class UserService {
     return builder;
   }
 
+  reactionsQueryBuilder(): SelectQueryBuilder<DbUserHighlightReaction> {
+    const builder = this.userHighlightReactionRepository.createQueryBuilder("user_highlight_reactions");
+
+    return builder;
+  }
+
+  async findTopTenUsers(): Promise<DbTopUser[]> {
```
Comment (brandonroberts) [inline src/user/user.service.ts:36]:
```suggestion
  async findTopUsers(limit = 10): Promise<DbTopUser[]> {
```

File: src/user/user.service.ts [hunk 4]
```diff
@@ -23,6 +27,28 @@ export class UserService {
...
+    const queryBuilder = this.reactionsQueryBuilder();
+
+    queryBuilder
+      .select("users.login as login")
+      .innerJoin("users", "users", "users.id = user_highlight_reactions.user_id")
+      .where("user_highlight_reactions.deleted_at IS NULL")
+      .groupBy("login")
+      .orderBy("COUNT(user_highlight_reactions.emoji_id)", "DESC")
+      .limit(10);
```
Comment (brandonroberts) [inline src/user/user.service.ts:45]:
```suggestion
      .limit(limit);
```

**Review by brandonroberts — CHANGES_REQUESTED** [review]

**Comment by brandonroberts** [comment]
> I'm not sure I got this right 🫠🫠. I'm open to criticism on this one 😄 cc: @brandonroberts

Pretty close 🙂

File: src/user/user.controller.ts [hunk 5]
```diff
@@ -82,4 +83,14 @@ export class UserController {
...
     return this.repoService.findAll(pageOptionsDto, user.id, ["TopRepos"]);
   }
+
+  @Get("/top")
+  @ApiOperation({
+    operationId: "getTop10Highlights",
+    summary: "List top users",
+  })
+  @ApiOkResponse({ type: DbTopUser })
+  async getTop10Highlights(@Query() limit: number): Promise<DbTopUser[]> {
```
Comment (brandonroberts) [inline src/user/user.controller.ts:93]: Let's remove this for now

```suggestion
  async getTop10Highlights(): Promise<DbTopUser[]> {
```

File: src/user/user.controller.ts [hunk 6]
```diff
@@ -82,4 +83,14 @@ export class UserController {
...
+    return this.userService.findTopTenUsers(limit);
```
Comment (brandonroberts) [inline src/user/user.controller.ts:94]: Use the default

```suggestion
    return this.userService.findTopUsers();
```

File: src/user/user.service.ts [hunk 7]
```diff
@@ -23,6 +27,28 @@ export class UserService {
     return builder;
   }
 
+  reactionsQueryBuilder(): SelectQueryBuilder<DbUserHighlightReaction> {
+    const builder = this.userHighlightReactionRepository.createQueryBuilder("user_highlight_reactions");
+
+    return builder;
+  }
+
+  async findTopTenUsers(limit = 10): Promise<DbTopUser[]> {
```
Comment (brandonroberts) [inline src/user/user.service.ts:36]:
```suggestion
  async findTopUsers(limit = 10): Promise<DbTopUser[]> {
```

**Review by brandonroberts — CHANGES_REQUESTED** [review]

**Review by brandonroberts — APPROVED** [review]

**Pull request merged**